aptos-move/aptos-vm/src/verifier/event_validation.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::move_vm_ext::SessionExt;
+use aptos_framework::RuntimeModuleMetadataV1;
+use move_binary_format::{
+    access::ModuleAccess,
+    errors::{Location, PartialVMError, VMError, VMResult},
+    file_format::{
+        Bytecode,
+        SignatureToken::{Struct, StructInstantiation},
+    },
+    CompiledModule,
+};
+use move_core_types::{
+    account_address::AccountAddress, language_storage::ModuleId, vm_status::StatusCode,
+};
+use std::collections::HashSet;
+
+const EVENT_MODULE_NAME: &str = "event";
+const EVENT_EMIT_FUNCTION_NAME: &str = "emit";
+
+fn metadata_validation_err(msg: &str) -> Result<(), VMError> {
+    Err(metadata_validation_error(msg))
+}
+
+fn metadata_validation_error(msg: &str) -> VMError {
+    PartialVMError::new(StatusCode::CONSTRAINT_NOT_SATISFIED)
+        .with_message(format!("metadata and code bundle mismatch: {}", msg))
+        .finish(Location::Undefined)
+}
+
+/// Validate resource group metadata on modules one by one:
+/// * Extract the event metadata
+/// * Verify all changes are compatible upgrades (existing event attributes cannot be removed)
+pub(crate) fn validate_module_events(
```

Comment:
cc @brianrmurphy I think this new attribute might need to be added somewhere in the compiler so it doesn't generate warnings for unknown attributes

Reply:
I fixed that